File: backend/app/schemas/audit.py
from __future__ import annotations
from datetime import datetime
from typing import Any
from app.models.enums import AuditActionEnum, AuditEntityTypeEnum
from .common import BaseSchema, IdSchema
from .project import ProjectRead
from .user import UserShortRead

class AuditLogCreate(BaseSchema):
    user_id: int | None = None
    project_id: int | None = None
    action: AuditActionEnum
    entity_type: AuditEntityTypeEnum
    entity_id: int | None = None
    before_json: dict[str, Any] | None = None
    after_json: dict[str, Any] | None = None
    ip_address: str | None = None
    user_agent: str | None = None

class AuditLogRead(IdSchema):
    user_id: int | None = None
    project_id: int | None = None
    action: AuditActionEnum
    entity_type: AuditEntityTypeEnum
    entity_id: int | None = None
    before_json: dict[str, Any] | None = None
    after_json: dict[str, Any] | None = None
    ip_address: str | None = None
    user_agent: str | None = None
    created_at: datetime

class AuditLogDetailRead(AuditLogRead):
    user: UserShortRead | None = None
    project: ProjectRead | None = None

File: frontend/src/shared/types/audit.ts
export type AuditAction =
  | 'create'
  | 'update'
  | 'delete'
  | 'submit'
  | 'approve'
  | 'reject'
  | 'process_start'
  | 'process_retry'
  | 'process_finish'
  | 'login'
  | 'logout'
  | 'export';

export type AuditEntityType =
  | 'user'
  | 'project'
  | 'report'
  | 'report_upload'
  | 'template'
  | 'task'
  | 'dashboard';

export interface AuditLog {
  id: number;
  user_id: number | null;
  project_id: number | null;
  action: AuditAction;
  entity_type: AuditEntityType;
  entity_id: number | null;
  before_json: Record<string, unknown> | null;
  after_json: Record<string, unknown> | null;
  ip_address: string | null;
  user_agent: string | null;
  created_at: string;
}

export interface AuditUserSummary {
  id: number;
  email: string;
  full_name: string;
  position: string | null;
  department: string | null;
  is_active: boolean;
}

export interface AuditProjectSummary {
  id: number;
  name: string;
  code: string;
  description: string | null;
  owner_id: number;
  is_archived: boolean;
  created_at: string;
  updated_at: string;
}

export interface AuditLogDetail extends AuditLog {
  user: AuditUserSummary | null;
  project: AuditProjectSummary | null;
}
